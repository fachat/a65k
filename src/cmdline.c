
/****************************************************************************

    commandline parsing
    Copyright (C) 2016-2017 Andre Fachat

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

****************************************************************************/

#include <stdlib.h>
#include <strings.h>
#include <string.h>

#include "hashmap.h"
#include "array_list.h"
#include "cmdline.h"
#include "err.h"
#include "infiles.h"

// hash from param name to cmdline_t sruct for quick find
static hash_t *params = NULL;
// list of parameters for help output
static list_t *paramlist = NULL;

static const char *prg_name = NULL;

static err_t usage(int flag, void* extra) {
	printf("Usage: %s [options] in-filename1 [in-filename2 ...]\n"
		"Cross-assembler for 65xx/R65C02/65816/65CE02/65002\n"
		"\n", prg_name); 

	list_iterator_t *iter = list_iterator(paramlist);
	cmdline_t *param = NULL;
	while ( (param = list_iterator_next(iter)) ) {
		printf("  -%s\n\t%s\n", param->name, param->description);
	}
	
	return E_ABORT;
}

static const cmdline_t help[] = {
	{ "?", PARTYPE_FLAG, NULL, usage, NULL, "Show this help" },
	{ "help", PARTYPE_FLAG, NULL, usage, NULL, "Show this help" },
};

static const char *key_from_param(const void *entry) {
	return ((cmdline_t*)entry)->name;
}

// TODO init the cmdline parser with the name as which the program was called
// for example "a65k", but also "xa65"
void cmdline_module_init() {

	params = hash_init_stringkey(50, 25, key_from_param);
	paramlist = array_list_init(20);

	cmdline_register_mult(help, sizeof(help)/sizeof(cmdline_t));
}

// template method where extra_param is the pointer to an int variable to set
err_t cmdline_set_flag(int flag, void *extra_param) {
	*((int*)extra_param) = flag;
	return E_OK;
}

void cmdline_register_mult(const cmdline_t *param, int num ) {
	for (int i = 0; i < num; i++) {
		cmdline_register(&param[i]);
	}
}

void cmdline_register(const cmdline_t *param) {
	
	if (hash_put(params, (void*)param) != NULL) {
		// must not happen
		exit(1);
	}
	list_add(paramlist, (void*)param);
}

err_t cmdline_parse(int argc, char *argv[]) {

	err_t rv = E_OK;

	prg_name = argv[0];

        int i = 1;
        while (i < argc && !rv) {
                if (argv[i][0] == '-') {
			char *val = NULL;
			int flag = 1;
			const char *name = argv[i]+1;
			if (name == strstr(name, "no-")) {
				name = name+3;
				flag = 0;
			}
			char *end = index(name, '=');
			if (end != NULL) {
				val = end + 1;
				end[0] = 0;
			}
			
			// lookup option
			cmdline_t *opt = hash_get(params, name);
			if (opt != NULL) {
				switch (opt->type) {
				case PARTYPE_FLAG:
					rv = opt->setflag(flag, opt->extra_param);
					break;
				case PARTYPE_PARAM:
					// TODO: error if param is missing
					// TODO: if started with "no-" return error
					rv = opt->setfunc(val, opt->extra_param);
					break;
				}
			} else {
				// TODO error cmdline parameter not found
			}
                } else {
			// register file with snapshot (clone) of current parser configuration
                        infiles_register(argv[i], parser_current_config());
                }   
		i++;
        }   
	return rv;
}




