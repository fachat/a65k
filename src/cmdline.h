
/****************************************************************************

    commandline parsing
    Copyright (C) 2016 Andre Fachat

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

****************************************************************************/


#ifndef CMDLINE_H
#define CMDLINE_H

#include "err.h"

// init the cmdline parser 
void cmdline_module_init();

// parse the options
err_t cmdline_parse(int argc, char *argv[]);

typedef enum {
	// no parameter, sets a bool when present with/without "no-" prefix
	PARTYPE_FLAG,
	// with required parameter
	PARTYPE_PARAM
} param_type_t;

typedef struct {
	// name of cmdline param
	const char 	*name;		

	// option has a parameter
	param_type_t	type;

	// function to call when parsed and type is PARTYPE_PARAM
	err_t 		(*setfunc)(const char *value, void *extra_param);

	// function to call when parsed and type is PARTYPE_FLAG
	void 		(*setflag)(int flag, void *extra_param);

	// extra param to be passed to set function
	void		*extra_param;

	// description
	const char	*description;
} cmdline_t;

// template method where extra_param is the pointer to an int variable to set
void cmdline_set_flag(int flag, void *extra_param);
 
void cmdline_register(const cmdline_t *param);

#endif

