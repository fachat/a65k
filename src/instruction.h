
/****************************************************************************

    CPU management
    Copyright (C) 2015 Andre Fachat

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

****************************************************************************/


#ifndef OPERATION_H
#define OPERATION_H



typedef enum {
	SY_IMP		= 0,	// implied
	SY_IMM		= 1,	// immediate
	
	SY_ABS		= 2,	// absolute/relative address: zp; word; long; quad; rel; relwide; rellong; relquad
	SY_ABSX		= 3,	// absolute x-indexed: zp,x; word,x; long,x; quad,x
	SY_ABSY		= 4,	// absolute y-indexed: zp,y; word,y; long,y; quad,y
	SY_ABSZ		= 5,	// absolute y-indexed: zp,y; word,y; long,y; quad,y

	SY_IND		= 6,	// zp/word-indirect, word address: (zp); (word)
	SY_INDY		= 7,	// zp/word-indirect, y-indexed, word address: (zp),y; (word),y
	SY_INDZ		= 8,	// zp/word-indirect, y-indexed, word address: (zp),y; (word),y
	SY_XIND		= 9,	// x-indexed, zp/word-indirect, word address: (zp,x); (word,x)

	SY_INDL		= 10,	// zp/word-indirect, long address: [zp]; [word]
	SY_INDYL	= 11,	// zp/word-indirect, y-indexed, long address: [zp],y; [word],y
	SY_INDZL	= 12,	// zp/word-indirect, y-indexed, long address: [zp],y; [word],y
	SY_XINDL	= 13,	// x-indexed, zp/word-indirect, long address: [zp,x]; [word,x]

	SY_INDQ		= 14,	// zp/word-indirect, quad address: [[zp]]; [[word]]
	SY_INDYQ	= 15,	// zp/word-indirect, y-indexed, quad address: [[zp]],y; [[word]],y
	SY_INDZQ	= 16,	// zp/word-indirect, y-indexed, quad address: [[zp]],y; [[word]],y
	SY_XINDQ	= 17,	// x-indexed, zp/word-indirect, quad address: [[zp,x]]; [[word,x]]
	
	SY_MV		= 18,	// two byte 65816 MVN/MVP addressing
	SY_BBREL	= 19,	// zeropage address, plus relative address, R65C02 BBR/BBS
	
	SY_MAX		= 20,	// end

	// virtual addressing modes. Will be mapped to SY_IND, SY_INDL or SY_INDQ depending
	// on memory model used during assembly
	SY_INDD		= 21,	// zp/word-indirect, dynamic address: ((zp)); ((word))
	SY_INDYD	= 22,	// zp/word-zp-indirect, y-indexed, dynamic address: ((zp)),y; ((word)),y
	SY_INDZD	= 23,	// zp/word-zp-indirect, y-indexed, dynamic address: ((zp)),y; ((word)),y
	SY_XINDD	= 24,	// x-indexed, zp/word-indirect, dynamic address: ((zp,x)); ((word,x))

} syntax_type;


typedef enum {
	AM_NONE		= -1,

	AM_IMP		= 0,	// implied
	AM_IMM8		= 1,	// immediate
	AM_IMM16	= 2,	// immediate
	AM_IMM32	= 3,	// immediate
	AM_IMM64	= 4,	// immediate
	
	AM_ABS8		= 5,	// absolute/relative address: zp
	AM_ABS16	= 6,	// absolute/relative address: word
	AM_ABS24	= 7,	// absolute address: word+bank byte
	AM_ABS32	= 8,	// absolute/relative address: long
	AM_ABS64	= 9,	// absolute/relative address: quad

	AM_ABS8X	= 10,	// absolute x-indexed: zp,x
	AM_ABS16X	= 11,	// absolute x-indexed: word,x
	AM_ABS32X	= 12,	// absolute x-indexed: long,x
	AM_ABS64X	= 13,	// absolute x-indexed: quad,x

	AM_ABS8Y	= 14,	// absolute y-indexed: zp,y
	AM_ABS16Y	= 15,	// absolute y-indexed: word,y
	AM_ABS32Y	= 16,	// absolute y-indexed: long,y
	AM_ABS64Y	= 17,	// absolute y-indexed: quad,y

	AM_IND8		= 18,	// zp/word-indirect, word address: (zp)
	AM_IND16	= 19,	// zp/word-indirect, word address: (word)
	AM_IND8Y	= 20,	// zp/word-indirect, y-indexed, word address: (zp),y
	AM_IND16Y	= 21,	// zp/word-indirect, y-indexed, word address: (word),y
	AM_IND8Z	= 22,	// zp/word-indirect, z-indexed, word address: (zp),z
	AM_IND16Z	= 23,	// zp/word-indirect, z-indexed, word address: (word),z
	AM_XIND8	= 24,	// x-indexed, zp/word-indirect, word address: (zp,x)
	AM_XIND16	= 25,	// x-indexed, zp/word-indirect, word address: (word,x)

	AM_IND8L	= 26,	// zp/word-indirect, long address: [zp]
	AM_IND16L	= 27,	// zp/word-indirect, long address: [word]
	AM_IND8YL	= 28,	// zp/word-indirect, y-indexed, long address: [zp],y
	AM_IND16YL	= 29,	// zp/word-indirect, y-indexed, long address: [word],y
	AM_IND8ZL	= 30,	// zp/word-indirect, z-indexed, long address: [zp],z
	AM_IND16ZL	= 31,	// zp/word-indirect, z-indexed, long address: [word],z
	AM_XIND8L	= 32,	// x-indexed, zp/word-indirect, long address: [zp,x]
	AM_XIND16L	= 33,	// x-indexed, zp/word-indirect, long address: [word,x]

	AM_IND8Q	= 34,	// zp/word-indirect, quad address: [[zp]]
	AM_IND16Q	= 35,	// zp/word-indirect, quad address: [[word]]
	AM_IND8YQ	= 36,	// zp/word-indirect, y-indexed, quad address: [[zp]],y
	AM_IND16YQ	= 37,	// zp/word-indirect, y-indexed, quad address: [[word]],y
	AM_IND8ZQ	= 38,	// zp/word-indirect, z-indexed, quad address: [[zp]],z
	AM_IND16ZQ	= 39,	// zp/word-indirect, z-indexed, quad address: [[word]],z
	AM_XIND8Q	= 40,	// x-indexed, zp/word-indirect, quad address: [[zp,x]]
	AM_XIND16Q	= 41,	// x-indexed, zp/word-indirect, quad address: [[word,x]]
	
	AM_MV		= 42,	// two byte 65816 MVN/MVP addressing
	AM_BBREL	= 43,	// zeropage address, plus relative address, R65C02 BBR/BBS
	
	AM_MAX		= 44,	// end
} amode_type;
	
typedef enum {
	PG_BASE		= 0,	// base page, no page prefix needed
	PG_EXT		= 1,	// extended page
	PG_QUICK	= 2,	// quick page
	PG_SYS		= 3,	// system page
} pg_type;

typedef enum {
	RS_NOT_SET	= 0,	// not set
        RS_BYTE         = 1,    // byte width 
        RS_WORD         = 2,    // word (2 byte) width
        RS_LONG         = 3,    // long (4 byte) width
        RS_QUAD         = 4     // quad (8 byte) width
} rs_type;

typedef enum {
	LE_NOT_SET	= 0,	// not set
        LE_E            = 1,    // no extension
        LE_S            = 2,    // sign extensions
        LE_0            = 3,    // extend with 0
        LE_1            = 4     // extend with 1
} le_type;


// ordering by SY_*, then AM_*
#define AMODES_MAX      5

typedef struct {
	const char	*pre;
	const char	*post;
	amode_type	amode[AMODES_MAX];
} syntax_t;

typedef struct {
	//bool_t		is_valid;
	unsigned char	code;
	pg_type 	page_prefix;
	const char 	*amodename;
	isa_map		isa;
} opcode_t;

// operation - equivalent to the mnemonic, like "lda", "adc", "inx", ...
typedef struct instruction_s instruction_t;

struct instruction_s {
	const char	*name;
	isa_map		isa;		// what ISA is it from? Also selects BCD and Illegal opcodes
	bool_t		abs_is_rel;	// is branch with relative addressing?
	bool_t		check_ac_w;	// check AC width mode for 65816, for adr modes with check_width
	bool_t		check_idx_w;	// check Index register width mode for 65816 (with check_width)
	instruction_t	*next;
	opcode_t	opcodes[AM_MAX];
};

typedef struct {
	bool_t		set_am_prefix;
	bool_t		rs_is_width;		// use RS-prefix for width
	pg_type		page_prefix;
	unsigned char	code;
} codepoint_t;

void instruction_module_init();

const instruction_t *instruction_find(const char *name);

bool_t opcode_find(const position_t *loc, const context_t *ctx, const instruction_t *op, syntax_type syntax, int opsize_in_bytes,
	codepoint_t *returned_code);

const syntax_t *op_syn_details(syntax_type syn);

inline static char prefix_um_char(bool_t um) {
	return um ? 'U' : '-';
}

inline static char prefix_nf_char(bool_t nf) {
	return nf ? 'N' : '-';
}

inline static char prefix_rs_char(rs_type p) {
	switch (p) {
	case RS_NOT_SET:
		return '-';
	case RS_BYTE:
		return 'B';
	case RS_WORD:
		return 'W';
	case RS_LONG:
		return 'L';
	case RS_QUAD:
		return 'Q';
	}
	return 'x';
}

inline static char prefix_le_char(rs_type p) {
	switch (p) {
	case LE_NOT_SET:
		return '-';
	case LE_0:
		return '0';
	case LE_1:
		return '1';
	case LE_S:
		return 'S';
	case LE_E:
		return 'E';
	}
	return 'x';
}

#endif

