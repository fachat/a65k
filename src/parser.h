/****************************************************************************

    parser
    Copyright (C) 2012 Andre Fachat

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

****************************************************************************/


#ifndef PARSER_H
#define PARSER_H

#include "context.h"
#include "label.h"
#include "block.h"
#include "operation.h"
#include "tokenizer.h"
#include "arith.h"
#include "list.h"

typedef enum {
        S_LABEQPC,              // set label to PC
        S_LABDEF,               // set label from parameter
	S_OPCODE,		// opcode
	S_PSEUDO,		// pseudo opcode
} stype_t;

/**
 * The statement_t struct is the main structure of the internal workings.
 * The parser creates a statement_t with all information to be gathered from the source.
 * The printer can "lint" this statement and pretty-format it
 * The assembler itself resolves labels and computes references and extends it with binary values
 * The printer can then also print the extended statement with binary values
 * The renderer can then put (a list of) statements into a binary format, depending on the output options
 */
typedef struct {
        const block_t           *blk;		// pointer to current block structure for label dereferencing
        const context_t         *ctx;		// current context, like CPU type, width, other options
        stype_t                 type;		// type of statement (see above)
        // optional
        const label_t           *label;		// TODO allow for multiple label definitions per line/statement
	// S_LABDEF
	const ilist_t		*setlabel;	// anode_t list for setting label to a value 
	// S_OPCODE
       	const operation_t       *op;		// pointer to info about operation, like "ADC", "TAX", ...
	op_t			base;		// base extension of the parameter, like ",B", ",s" if any, zero else
	
	syntax_type		syn;		// addressing mode syntax, see SY_* constants
       	ilist_t  	        *param;		// list of anode_t structs that describe the syntax tree for the parameter value
} statement_t;


void parser_module_init(void);

err_t parser_push(const context_t *context, const line_t *line);

list_iterator_t *parser_get_statements(void);

// for test cases only
void parser_reset();

#endif

