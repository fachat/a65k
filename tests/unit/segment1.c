

#include <stdio.h>

#include "types.h"
#include "position.h"
#include "cpu.h"
#include "segment.h"
#include "context.h"
#include "position.h"
#include "instruction.h"
#include "cmdline.h"
#include "parser.h"
#include "dbgprint.h"
#include "log.h"


static void test(const context_t *ctx, position_t *pos, const char *txt) {

	parser_reset();

	line_t line = { txt, pos, parser_current_config() };

	do_print("");
	do_print(">>> %s", line.line);
		
	err_t err = parser_push(ctx, &line);

	if (err) {
		do_print("err=%d", err);
	}
	//log_set_level(LEV_DEBUG);

	list_iterator_t *iter = parser_get_statements();
	while (list_iterator_has_next(iter)) {
		statement_t *stmt = list_iterator_next(iter);
		print_debug_stmt(stmt);
	}
	list_iterator_free(iter);
}


int main(int argc, char *argv[]) {

        (void)argc;
        (void)argv;

	cmdline_module_init();
	cpu_module_init();
	segment_module_init();
	operation_module_init();
	parser_module_init();
	tokenizer_module_init();

	//parser_config_init();

	const cpu_t *cpu = cpu_by_name("nmos");

	segment_t *seg = segment_new(NULL, "seg1", SEG_ANY, cpu->type, true);

	context_t *ctx = (context_t*) context_init(seg, cpu);

	position_t pos = { "bogusfile", 1 };

	test(ctx, &pos, "label adc #123");	

	ctx = context_dup();
	ctx->segment = segment_new(NULL, "seg1", SEG_ANY, cpu->type, true);

	test(ctx, &pos, "lda 123");	

	ctx = context_dup();
	ctx->segment = segment_new(NULL, "seg2", SEG_ANY, cpu->type, true);

	test(ctx, &pos, "ldx (12),y");	

	ctx = context_dup();
	ctx->segment = segment_new(NULL, "seg1", SEG_ANY, cpu->type, true);

	test(ctx, &pos, "lda ($12,s,x)");	

	ctx = context_dup();
	ctx->segment = segment_new(NULL, "seg2", SEG_ANY, cpu->type, true);

	test(ctx, &pos, "lda (12,s),y");	


	do_print("By segment");
	list_iterator_t *iter = segment_get_iter();
	while (list_iterator_has_next(iter)) {
		segment_t *segm = list_iterator_next(iter);

		do_print("got segment: %s", segm->name);

		list_iterator_t *stiter = segment_get_statements(segm);
		while(list_iterator_has_next(stiter)) {
			statement_t *stmt = list_iterator_next(stiter);
			print_debug_stmt(stmt);
		}
		list_iterator_free(stiter);
	}
	list_iterator_free(iter);
}



