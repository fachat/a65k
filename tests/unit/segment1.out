WRN:
WRN:>>> label adc #123
WRN:--------------- STMT ---------------
WRN:BLK:0 ctx:0 (CPU: nmos, isa:$7), seg=seg1
WRN:LAB: name:label ctx:0 pos:bogusfile:1
WRN:OPR: name:ADC isa:$561 isrel:0 acw:0 idxw:0
WRN:MOD: BASE: 0, SYN: 1, UM: 0, NF: 0, RS: 0, LE: 0
WRN:PAR: 
WRN:    A:len=1
WRN:      type=v, modifier=0 ( ), op=0( ), val=(d)123/7b
WRN:
WRN:>>> lda 123
WRN:--------------- STMT ---------------
WRN:BLK:0 ctx:1 (CPU: nmos, isa:$7), seg=seg1
WRN:OPR: name:LDA isa:$1561 isrel:0 acw:0 idxw:0
WRN:MOD: BASE: 0, SYN: 2, UM: 0, NF: 0, RS: 0, LE: 0
WRN:PAR: 
WRN:    A:len=1
WRN:      type=v, modifier=0 ( ), op=0( ), val=(d)123/7b
WRN:
WRN:>>> ldx (12),y
WRN:--------------- STMT ---------------
WRN:BLK:0 ctx:2 (CPU: nmos, isa:$7), seg=seg2
WRN:OPR: name:LDX isa:$401 isrel:0 acw:0 idxw:0
WRN:MOD: BASE: 0, SYN: 7, UM: 0, NF: 0, RS: 0, LE: 0
WRN:PAR: 
WRN:    A:len=1
WRN:      type=v, modifier=0 ( ), op=0( ), val=(d)12/c
WRN:
WRN:>>> lda ($12,s,x)
WRN:--------------- STMT ---------------
WRN:BLK:0 ctx:3 (CPU: nmos, isa:$7), seg=seg1
WRN:OPR: name:LDA isa:$1561 isrel:0 acw:0 idxw:0
WRN:MOD: BASE: 151, SYN: 9, UM: 0, NF: 0, RS: 0, LE: 0
WRN:PAR: 
WRN:    A:len=1
WRN:      type=v, modifier=0 ( ), op=0( ), val=(h)18/12
WRN:
WRN:>>> lda (12,s),y
WRN:--------------- STMT ---------------
WRN:BLK:0 ctx:4 (CPU: nmos, isa:$7), seg=seg2
WRN:OPR: name:LDA isa:$1561 isrel:0 acw:0 idxw:0
WRN:MOD: BASE: 151, SYN: 7, UM: 0, NF: 0, RS: 0, LE: 0
WRN:PAR: 
WRN:    A:len=1
WRN:      type=v, modifier=0 ( ), op=0( ), val=(d)12/c
WRN:By segment
WRN:got segment: seg1
WRN:--------------- STMT ---------------
WRN:BLK:0 ctx:0 (CPU: nmos, isa:$7), seg=seg1
WRN:LAB: name:label ctx:0 pos:bogusfile:1
WRN:OPR: name:ADC isa:$561 isrel:0 acw:0 idxw:0
WRN:MOD: BASE: 0, SYN: 1, UM: 0, NF: 0, RS: 0, LE: 0
WRN:PAR: 
WRN:    A:len=1
WRN:      type=v, modifier=0 ( ), op=0( ), val=(d)123/7b
WRN:--------------- STMT ---------------
WRN:BLK:0 ctx:1 (CPU: nmos, isa:$7), seg=seg1
WRN:OPR: name:LDA isa:$1561 isrel:0 acw:0 idxw:0
WRN:MOD: BASE: 0, SYN: 2, UM: 0, NF: 0, RS: 0, LE: 0
WRN:PAR: 
WRN:    A:len=1
WRN:      type=v, modifier=0 ( ), op=0( ), val=(d)123/7b
WRN:--------------- STMT ---------------
WRN:BLK:0 ctx:3 (CPU: nmos, isa:$7), seg=seg1
WRN:OPR: name:LDA isa:$1561 isrel:0 acw:0 idxw:0
WRN:MOD: BASE: 151, SYN: 9, UM: 0, NF: 0, RS: 0, LE: 0
WRN:PAR: 
WRN:    A:len=1
WRN:      type=v, modifier=0 ( ), op=0( ), val=(h)18/12
WRN:got segment: seg2
WRN:--------------- STMT ---------------
WRN:BLK:0 ctx:2 (CPU: nmos, isa:$7), seg=seg2
WRN:OPR: name:LDX isa:$401 isrel:0 acw:0 idxw:0
WRN:MOD: BASE: 0, SYN: 7, UM: 0, NF: 0, RS: 0, LE: 0
WRN:PAR: 
WRN:    A:len=1
WRN:      type=v, modifier=0 ( ), op=0( ), val=(d)12/c
WRN:--------------- STMT ---------------
WRN:BLK:0 ctx:4 (CPU: nmos, isa:$7), seg=seg2
WRN:OPR: name:LDA isa:$1561 isrel:0 acw:0 idxw:0
WRN:MOD: BASE: 151, SYN: 7, UM: 0, NF: 0, RS: 0, LE: 0
WRN:PAR: 
WRN:    A:len=1
WRN:      type=v, modifier=0 ( ), op=0( ), val=(d)12/c
